spec/acceptance/trusted_external_data_spec.rb

```diff
@@ -1,26 +1,114 @@
 
-context 'trusted external data' do
-  context '$trusted.external.servicenow hash' do
-    before(:all) do
-      manifest = <<-MANIFEST
+describe 'trusted external data ($trusted.external.servicenow hash)' do
+  let(:params) do
+    servicenow_config = servicenow_instance.bolt_config['remote']
+
+    {
+      instance: servicenow_instance.uri,
+      user: servicenow_config['user'],
+      password: servicenow_config['password'],
+    }
+  end
+  let(:setup_manifest) do
+    to_manifest(declare('Service', 'pe-puppetserver'), declare('class', 'servicenow_cmdb_integration', params))
+  end
+  let(:cmdb_table) { nil }
+  let(:certname_field) { 'fqdn' }
+
+  before(:all) do
+    manifest = <<-MANIFEST
       $trusted_json = inline_template("<%= @trusted.to_json %>")
       notify { "trusted external data":
         message => "#{TRUSTED_JSON_SEPARATOR}${trusted_json}#{TRUSTED_JSON_SEPARATOR}"
       }
-      MANIFEST
-      set_sitepp_content(manifest)
-    end
+    MANIFEST
+    set_sitepp_content(manifest)
+  end
+  after(:all) do
+    set_sitepp_content('')
+  end
+
+  # Tests may have test-specific parameters for the servicenow_cmdb_integration
+  # class so make sure to apply the class _before_ each test
+  before(:each) do
+    master.apply_manifest(setup_manifest)
+  end
 
-    after(:all) do
-      set_sitepp_content('')
+  it 'has idempotent setup' do
+    master.apply_manifest(setup_manifest, catch_changes: true)
+  end
+
+  shared_context 'setup cmdb' do |cmdb_table = 'cmdb_ci', certname_field = 'fqdn'|
+    before(:each) do
+      fields_template = JSON.parse(File.read('spec/support/acceptance/cmdb_record_template.json'))
+      # Store the CMDB table in an arbitrary (String) field so that tests can assert on it
+      fields_template['attributes'] = cmdb_table
+      CMDBHelpers.create_target_record(master, fields_template, table: cmdb_table, certname_field: certname_field)
+    end
+    after(:each) do
+      CMDBHelpers.delete_target_record(master, table: cmdb_table, certname_field: certname_field)
     end
+  end
+
+  context 'default behavior' do
+    include_context 'setup cmdb'
 
-    it "contains the node's CMDB record" do
+    it "contains the node's CMDB record in the 'cmdb_ci' table obtained by querying the 'fqdn' field" do
       result = trigger_puppet_run(master)
       trusted_json = parse_trusted_json(result.stdout)
       cmdb_record = CMDBHelpers.get_target_record(master)
       expect(trusted_json['external']['servicenow']).to eql(cmdb_record)
     end
   end
+
+  context 'user specifies a different CMDB table' do
+    let(:params) { super().merge('table' => 'cmdb_ci_acc') }
+
+    # NOTE: cmdb_ci_acc is an arbitrary but real table on an actual ServiceNow instance.
+    # If for some reason it is changed to something else, then make sure to update the
+    # mock ServiceNow instance's implementation with the new table since the mock's
+    # available tables are (currently) hardcoded.
```

Comment:
All ServiceNow tables are tracked in a separate table (`sys_db_object`). So the way to "automate" this is to list that `sys_db` table's records and pick a random record derived from `cmdb_ci` -- that random record's `name` field will be the "different table" that we can use in this test. Then the mock ServiceNow instance won't have to worry about hardcoding table names, only that its other table is "derived" from `cmdb_ci`.

I'd like to do that improvement later since it isn't that important, but thought I'd talk about it here.